### PR TITLE
feat: AWS SecretsManager Config (allow ForceDeleteWithoutRecovery for PushSecret)

### PR DESCRIPTION
**Pull request by korenyoni**
## Problem Statement

Allow [ForceDeleteWithoutRecovery](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DeleteSecret.html#API_DeleteSecret_RequestSyntax) and [RecoveryWindowInDays](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DeleteSecret.html#API_DeleteSecret_RequestSyntax) in PushSecret deletion for AWS SecretsManager via a new `spec.secretsManager` field in `SecretStore`.

## Related Issue

Closes https://github.com/external-secrets/external-secrets/issues/2830 

## Proposed Changes

This PR adds a new `spec.secretsManager` field in `SecretStore` that allows for setting [ForceDeleteWithoutRecovery](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DeleteSecret.html#API_DeleteSecret_RequestSyntax) and [RecoveryWindowInDays](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DeleteSecret.html#API_DeleteSecret_RequestSyntax) (these fields are mutually exclusive).

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

## Notes

Aside from extending the test suite, I tested this by following [development guide](https://external-secrets.io/latest/contributing/devguide/#installing) on running the custom ESO build on a kind cluster.

<details>
<summary>ForceDeleteWithoutRecovery</summary>
  
```yaml
apiVersion: external-secrets.io/v1beta1
kind: SecretStore
metadata:
  name: team-b-store
spec:
  provider:
    aws:
      service: SecretsManager
      region: us-east-1
      auth:
        secretRef:
          accessKeyIDSecretRef:
            name: aws-rnd-sandbox
            key: access-key
          secretAccessKeySecretRef:
            name: aws-rnd-sandbox
            key: secret-access-key
      secretsManager:
        forceDeleteWithoutRecovery: true
        #recoveryWindowInDays: 9
```

(secret was deleted immediately)

</details>

<details>
<summary>RecoveryWindowInDays</summary>
  
```yaml
apiVersion: external-secrets.io/v1beta1
kind: SecretStore
metadata:
  name: team-b-store
spec:
  provider:
    aws:
      service: SecretsManager
      region: us-east-1
      auth:
        secretRef:
          accessKeyIDSecretRef:
            name: aws-rnd-sandbox
            key: access-key
          secretAccessKeySecretRef:
            name: aws-rnd-sandbox
            key: secret-access-key
      secretsManager:
        #forceDeleteWithoutRecovery: true
        recoveryWindowInDays: 9
```

CleanShot 2023-11-06 at 10 25 30

(neither the UI nor the CLI show the recovery window, but the AWS Secrets Manager behaviour seems to match the expected behavior)

</details>

### Misc changes:
* Updated the description for `ClusterSecretStore`/`SecretStore` attributes that are relevant to both AWS Secrets Manager and AWS Parameter Store providers, but refer only to the former. For example: https://github.com/external-secrets/external-secrets/blob/a9d70527b3a8f7edb03557c817f3512d4b43a52a/apis/externalsecrets/v1beta1/secretstore_aws_types.go#L81
* Fix grammatical errors in `ClusterSecretStore`/`SecretStore` attribute descriptions
* Remove duplicate AWS Provider test: https://github.com/external-secrets/external-secrets/blob/a9d70527b3a8f7edb03557c817f3512d4b43a52a/pkg/provider/aws/provider_test.go#L185-L213
* Fix apparent copy-paste issue in test suite logging (refers to `SetSecret` when actually testing `DeleteSecret`).

### CODE REVIEW

**Comment by Skarlso** [comment]
Errr, something might have happend during rebase there?

**Comment by gusfcarvalho** [comment]
Hi!

originally we thought of having specific PushSecret configuration under a structure of its own on the SecretStore (see https://github.com/external-secrets/external-secrets/blob/main/design/002-pushsecret.md). At the time we envisioned only encryption configuration, but I guess it would make sense to package it all under `pushConfig` or something similar.

**Comment by korenyoni** [comment]
> Hi!
> 
> originally we thought of having specific PushSecret configuration under a structure of its own on the SecretStore (see https://github.com/external-secrets/external-secrets/blob/main/design/002-pushsecret.md). At the time we envisioned only encryption configuration, but I guess it would make sense to package it all under `pushConfig` or something similar (in this case, under spec.provider.AWS)

Yeah, I had to think a bit about what would be the best place to place these settings in. 

`pushConfig` would work as well, but either way you will need a substructure that's specific to AWS SecretsManager, so it would need to be `pushConfig.secretsManager`. I'm open to moving these settings somewhere else, if needed.

File: apis/externalsecrets/v1beta1/secretstore_aws_types.go [hunk 1]
```diff
@@ -54,14 +54,24 @@ type AWSJWTAuth struct {
...
 )
 
+// SecretsManager defines how the provider behaves when interacting with AWS
+// SecretsManager. Some of these settings are only applicable to controlling how
+// secrets are deleted, and hence only apply to PushSecret.
+type SecretsManager struct {
+	// +optional
+	ForceDeleteWithoutRecovery bool `json:"forceDeleteWithoutRecovery,omitempty"`
+	// +optional
+	RecoveryWindowInDays int64 `json:"recoveryWindowInDays,omitempty"`
```
Comment (moolen) [inline apis/externalsecrets/v1beta1/secretstore_aws_types.go:72]: Can you please add a comment for these options (what they do, that they apply only to PushSecrets, valid ranges for recovery window, also consider pointing to official AWS docs), that'll end up in the openAPI spec and will help users to explore these options e.g. with `kubectl explain` or other means.

(you have that already below in the example `SecretStore` in the documentation)

**Review by moolen — COMMENTED** [review]
The proposed CRD changes look good to me. I think it's fine if the properties are documented so it is clear for the user where these apply to (ExternalSecret/PushSecret, both...).

Not sure about `pushConfig`, i don't have a strong opinion on that. I think it would be nice, but also a bit inconsistent. Some of the existing properties apply only to ExternalSecrets, others are shared (among AWS Secrets Manager / Parameter Store..). I feel there's already quite a complex structure due to reusing properties across AWS SM/PS.

File: apis/externalsecrets/v1beta1/secretstore_aws_types.go (hunk 1 above)
Comment (korenyoni) [reply apis/externalsecrets/v1beta1/secretstore_aws_types.go]: Done!

I ended up grabbing the descriptions from the AWS API docs. `ForceDeleteWithoutRecovery` is extremely long, so I just added the first paragraph (and linked the docs for both fields).

**Review by moolen — APPROVED** [review]
LGTM :cat: 

@gusfcarvalho any objections regarding the CRD structure, should we hold this off and further discuss it?

**Pull request merged**